Main.hs

```diff
@@ -113,100 +113,169 @@ hsQName builtins f
+    Con c _ es
+      | []      <- vis es, conName c ~~ nil  -> return []
+      | [x, xs] <- vis es, conName c ~~ cons -> (x :) <$> makeList' nil cons err xs
+    _ -> genericDocError err
+  where
+    vis es = [ unArg a | Apply a <- es, visible a ]
+
+makeListP' :: String -> String -> Doc -> DeBruijnPattern -> TCM [DeBruijnPattern]
+makeListP' nil cons err p = do
+  case p of
+    ConP c _ ps
+      | []      <- vis ps, conName c ~~ nil  -> return []
+      | [x, xs] <- vis ps, conName c ~~ cons -> (x :) <$> makeListP' nil cons err xs
+    _ -> genericDocError err
   where
-    (|->) = (,)
-    getB (b, t) = getBuiltin' b >>= \ case
-      Nothing          -> return []
-      Just (Def q _)   -> return [(q, t)]
-      Just (Con c _ _) -> return [(conName c, t)]
-      Just _           -> __IMPOSSIBLE__
-
-compilePrim :: Prim -> Hs.QName ()
-compilePrim = \case
-  Unit   -> special Hs.UnitCon
-  Nat    -> unqual "Integer"
-  Float  -> unqual "Double"
-  Word64 -> unqual "Word64"
-  Bool   -> unqual "Bool"
-  True_  -> unqual "True"
-  False_ -> unqual "False"
-  Char   -> unqual "Char"
-  List   -> special Hs.ListCon
-  Cons   -> special Hs.Cons
-  Nil    -> special Hs.ListCon
+    vis ps = [ namedArg p | p <- ps, visible p ]
+
+underAbstr :: Subst t a => Dom Type -> Abs a -> (a -> TCM b) -> TCM b
+underAbstr a b ret = underAbstraction' KeepNames a b $ \ body ->
+  localScope $ bindVar 0 >> ret body
+
+underAbstr_ :: Subst t a => Abs a -> (a -> TCM b) -> TCM b
+underAbstr_ = underAbstr __DUMMY_DOM__
+
+-- Builtins ---------------------------------------------------------------
+
+isSpecialTerm :: QName -> Maybe (QName -> Elims -> TCM (Hs.Exp ()))
+isSpecialTerm = show >>> \ case
+  "Haskell.Prelude.if_then_else_" -> Just ifThenElse
+  _ -> Nothing
+
+isSpecialCon :: QName -> Maybe (ConHead -> ConInfo -> Elims -> TCM (Hs.Exp ()))
+isSpecialCon = show >>> \ case
+  "Haskell.Prelude.Tuple._∷_" -> Just tupleTerm
+  _ -> Nothing
+
+isSpecialPat :: QName -> Maybe (ConHead -> ConPatternInfo -> [NamedArg DeBruijnPattern] -> TCM (Hs.Pat ()))
+isSpecialPat = show >>> \ case
+  "Haskell.Prelude.Tuple._∷_" -> Just tuplePat
+  _ -> Nothing
+
+isSpecialType :: QName -> Maybe (QName -> Elims -> TCM (Hs.Type ()))
+isSpecialType = show >>> \ case
+  "Haskell.Prelude.Tuple" -> Just tupleType
+  _ -> Nothing
+
+isSpecialName :: QName -> Maybe (Hs.QName ())
+isSpecialName = show >>> \ case
+    "Agda.Builtin.Nat.Nat"         -> unqual "Integer"
+    "Agda.Builtin.Float.Float"     -> unqual "Double"
+    "Agda.Builtin.Bool.Bool.false" -> unqual "False"
+    "Agda.Builtin.Bool.Bool.true"  -> unqual "True"
+    "Agda.Builtin.List.List"       -> special Hs.ListCon
+    "Agda.Builtin.List.List._∷_"   -> special Hs.Cons
+    "Agda.Builtin.List.List.[]"    -> special Hs.ListCon
+    "Agda.Builtin.Unit.⊤"          -> special Hs.UnitCon
+    "Agda.Builtin.Unit.tt"         -> special Hs.UnitCon
+    "Haskell.Prelude.Tuple.[]"     -> special Hs.UnitCon
+    _ -> Nothing
```

Comment:
Fantastic, that's much cleaner to work with and no need for the dummy intermediate `Prim` datatype!